12229 read outside event loop

jasobrown opened the pull request.

jasobrown committed:
switch internode messaging to netty

patch by jasobrown, reviewed by pcmanus for CASSANDRA-8457

jasobrown committed:
move streaming to use netty

patch by jasobrown, reviewed by XYZ for CASSANDRA-12229

jasobrown committed:
adding circleci support

jasobrown committed:
- changed sending of stream control messages. now they are sent on the same (non-blocking) channels/sockets as the the actual file transfer. This replaces the earlier behavior of using the MessagingService for those messages.
- added KeepAlive back in
- minor clean up in the message classes
- removed streaming message classes from MS.Verb

jasobrown committed:
add explicit protocolVersion to OutboundConfigParams

jasobrown committed:
more fixes for the bidirectional messaging

jasobrown committed:
pull in updated circle config from trunk

jasobrown committed:
fixed bug on crc check

jasobrown committed:
fix data race of releasing the sstable ref

jasobrown committed:
remove local-test entries for flight recorder

jasobrown committed:
WIP - move disk IO outside the netty event loop, and bring back a lot of the originl streaming code

jasobrown committed:
WIP - fixing bugs

jasobrown committed:
test clean up

jasobrown committed:
unhack yamls

jasobrown committed:
import clean up; revert StreamRateLimiter; reintroduce CompressedInputStreamTest

jasobrown committed:
got this branch working

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufReadableByteChannel.java:
```diff
@@ -0,0 +1,195 @@
+     */
+    private int updateCurrentBuffer() throws EOFException
+    {
+        if (buffer != null)
```
Don't allow null buffers use a static final zero length buffer so you can remove all null checks?

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufReadableByteChannel.java:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * The count of readable bytes in all {@link ByteBuf}s held by this instance.
+     */
+    private volatile int queuedByteCount;
```
My thread safety concerns still exist. I also realized that they can stomp on each other's updates completely changing the value so it decrements or increments are completely lost.

> This would be incremented by one thread and decremented by another?
>
>Additionally incrementing is not atomic with waking up the consumer so the value could go negative until the increment occurs. I would synchronize append and updateBuffer and call it good. You shouldn't enter update buffer unless you are actually draining a buffer and buffers should be relatively large.

aweisberg replied:
You could also just make it atomic.

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufReadableByteChannel.java:
```diff
@@ -0,0 +1,195 @@
+     * When a {@link ByteBuf} is pulled off the {@link #queue}, capture it's readable byte count before we start reading
+     * from it, that we we can properly decrement {@link #queuedByteCount} when we're done with the buffer.
+     */
+    private int bufferInitialReadableBytes;
```
This is unused and I still think it wouldn't make it thread safe.

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufReadableByteChannel.java:
```diff
@@ -0,0 +1,195 @@
+                throw new EOFException();
+            try
+            {
+                buffer = queue.poll(1, TimeUnit.SECONDS);
```
I know it's picky, but can you put a sentinel value in the queue and use take and wake it up immediately? Maybe the empty zero length ByteBuf?

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufReadableByteChannel.java:
```diff
@@ -0,0 +1,195 @@
+import io.netty.util.ReferenceCountUtil;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class ByteBufReadableByteChannel implements ReadableByteChannel
```
So the bad news here is that this is going to introduce overhead. Channels.newChannel + the wrapper data input stream are going to read everything out byte at a time in the worst possible way. Wrapping ByteBuf as an InputStream was pretty good. Most optimal would be having a way to extend RebufferingInputStream to access the ByteBuf via a ByteBuffer wrapper.

Having fewer wrappers is better which is why translating a pipeline into a something that pushes buffers up to RebufferingInputStream is the fastest.

Can you avoid having this as a channel and have it just be a queue of buffers that manages high and low watermarks?

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufReadableByteChannel.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public void close() throws IOException
```
Which threads can call close? Is it thread safe if the reader is reading? The reader could see buffers out of order since the closer is also closing releasing buffers at the same time the reader is polling?

aweisberg commented on src/java/org/apache/cassandra/net/async/NettyFactory.java:
```diff
@@ -307,5 +322,35 @@ public void close()
+
+    private static Checksum createXXHashChecksum()
+    {
+        return XXHashFactory.fastestInstance().newStreamingHash32(XXHASH_DEFAULT_SEED).asChecksum();
```
You almost always want the 64-bit checksum because it's faster to calculate. http://cyan4973.github.io/xxHash/ 
Wider hashes up to 16-bytes can use the two 8-byte integer ports to calculate the hash in parallel.

aweisberg replied:
Now the real question is why not use the CRC32 hardware instruction? A CRC is guaranteed to detect a single bit flip and a hash isn't, but with 64-bits of output maybe it's not so important.

The CRC32 intrinsic also might be faster because it will handle on heap and off heap memory correctly without making a copy. The instance you picked is going to do a JNI call and copy all the bytes on every call. JNI copying overhead is a meaningful penalty and avoiding it by doing things in Java or by always using direct byte buffers pays off for real.

aweisberg replied:
Any thoughts on this?

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+
+    boolean checkBufSpace(int writeSize)
+    {
+        if (buf != null)
```
Why lazily allocate this? In what situation would you not have a buffer? I think you should always start with the canonical zero length ByteBuf and let it flush and write it. Assuming it's safe to flush a shared 0 length buffer (who knows!).

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    {
+        if (buf != null)
+        {
+            if (buf.writableBytes() >= writeSize)
```
Can write size be > buffer capacity?

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void write(int v) throws IOException
+    {
+        checkBufSpace(Byte.BYTES);
```
You need to manually inline a check for if(buf.writableBytes() > 0) into all the write methods. It definitely should be there since it has to be done every single time and the rest of the method should be separate because it runs infrequently and doesn't necessarily need to be inlined.

If you push all the checks into the method then the inlining options are much more limited because the compiler won't inline partial methods. So either you end up with extra function calls or a bunch of extra inlined code you don't need. This limits inlining options further up the call stack since you allocated a bunch of instructions to inline code that you really didn't want to.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void write(ByteBuffer byteBuffer) throws IOException
+    {
+        final boolean onHeap = byteBuffer.hasArray();
```
Doesn't Netty already have a method you can delegate to for writing a ByteBuffer to a ByteBuf that do the right thing with direct buffers?

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+        {
+            int size = src.remaining();
+            // This is unfortunate: I can't wrap a nio direct buffer with a netty wrapper -
+            // the buffer gets corrupted on the transfer. so for now, just copying into a ByteBuf
```
Is this a Netty bug?

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void writeFloat(float v) throws IOException
+    {
+        writeInt(Float.floatToIntBits(v));
```
You are not using the raw method here which we use in the other streams? Which is the correct one to use?

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void writeDouble(double v) throws IOException
+    {
+        writeLong(Double.doubleToLongBits(v));
```
Same raw question

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    public void writeByte(int v) throws IOException
+    {
+        checkBufSpace(Byte.BYTES);
+        buf.writeByte((byte) (v & 0xFF));
```
You don't need to mask on narrowing conversions only widening conversions which will do sign extension.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    public void write(int v) throws IOException
+    {
+        checkBufSpace(Byte.BYTES);
+        buf.writeByte((byte) (v & 0xFF));
```
You don't need to mask.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    public void writeBytes(String s) throws IOException
+    {
+        for (int index = 0; index < s.length(); index++)
+            writeByte(s.charAt(index) & 0xFF);
```
Mask isn't necessary. The narrowing conversion will take care of it.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+        checkBufSpace(size);
+        if (size == 1)
+        {
+            buf.writeByte((byte) (v & 0xFF));
```
Don't need the mask the narrowing conversion will take care of it.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void close()
+    {
+        if (channel.isActive())
```
If you flush and the channel isn't active since it's a race what happens?

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamCompressionInputStream.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public int read() throws IOException
+    {
```
This implementation is going to be slow to read from for primitives because it's going to pull everything in byte at a time with all the redundant bounds checking and so on.

We did a lot of work to get away from streams wrapping streams and to reduce it to just the bounds check against a ByteBuffer and then the read with a slow path if the bounds check failed to pull the next ByteBuffer.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java:
```diff
@@ -0,0 +1,105 @@
+        final int compressedLength = headerBuffer.getInt();
+        final int uncompressedLength = headerBuffer.getInt();
+
+        ByteBuffer compressed = ByteBuffer.allocateDirect(compressedLength);
```
We know that in practice the channel is backed by a queue of ByteBufs. We should really avoid this copy and not work with a ReadableByteChannel or cheat and cast to a known type that will just give us the ByteBuf.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+ * and be assocaited with the promises of the buffers into which it was serialized. Several messages may be packed into
+ * the same buffer, so each of those messages will share the same promise.
+ */
+public class SwappingByteBufDataOutputStreamPlus extends DataOutputStreamPlus
```
This might be better implemented as a WritableByteChannel and then use an extension of BufferedDataOutputStreamPlus to manage creating allocating and flushing ByteBufs/ByteBuffers? You can recycle the same ByteBuffer object and reassign the internal values with the contents of a ByteBuf. Unless Netty is going to screw you over on this.

I think there is a lot of redundant functionality here that isn't plugging into the existing stream classes we have that work with ByteBuffers. They are designed to be extended to work with this stuff but still use common methods as much as possible.

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamCompressionInputStream.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.cassandra.streaming.async.StreamCompressionSerializer;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class StreamCompressionInputStream extends InputStream
```
I think this class should definitely extend RebufferingInputStream and implement rebuffer.

aweisberg replied:
The situation with TrackedDataInputStream and the whole wrapped stack is suboptimal. We wait until to late in the process to switch to DataInputPlus and lose the ability to do efficient reads of primitives instead going down the byte at a time path for reads.

Granted this sub-optimal situation predates your current streaming code.

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -627,25 +610,39 @@ public void prepare(Collection<StreamRequest> requests, Collection<StreamSummary
      *
      * @param header sent header
      */
+    // TODO:JEB should this be called?????
```
Now it is called so can we remove this?

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -530,29 +523,48 @@ public boolean isSuccess()
+                    receive((PrepareAckMessage) message);
+                    break;
+                case FILE:
+                    receive((IncomingFileMessage) message);
```
I'm a little confused because it sure looks like we never actually transfer the file before getting to here? The answer of course is that it's tucked away in the IncomingFileMessageSerializer. Probably work a comment.

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamReader.java:
```diff
@@ -99,9 +98,12 @@ public SSTableMultiWriter read(ReadableByteChannel channel) throws IOException
+        TrackedInputStream in = new TrackedInputStream(new StreamCompressionInputStream(channel, StreamSession.CURRENT_VERSION, compressionSerializer));
         StreamDeserializer deserializer = new StreamDeserializer(cfs.metadata(), in, inputVersion, getHeader(cfs.metadata()));
         SSTableMultiWriter writer = null;
+        int cnt = 0;
```
cnt is unused now?

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamWriter.java:
```diff
@@ -131,28 +139,95 @@ protected long totalSize()
+        }
+        catch (InterruptedException e)
+        {
+            // TODO:JEB handle this better
```
Ignoring interrupted exception is almost always wrong. Either set interrupted again and return, throw some other exception, or handle the reason why you were interrupted.

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamWriter.java:
```diff
@@ -131,28 +139,95 @@ protected long totalSize()
+        // wait until the channel is writable again.
+        SimpleCondition condition = new SimpleCondition();
+        ChannelHandler writabilityHandler = new ChannelWritabiliityListsener(condition);
+        channel.pipeline().addLast(writabilityHandler);
```
This is safely atomic with changes to writability? No possibility of lost wakeup? This just seems a bit convoluted as an idiom on Netty's part. Seems like you should be able to block on flush.

Do you need to create the listener and then check writability again before blocking?

It also seems like there is a race here because Netty might not update writability immediately, but here we are pushing as fast we can? Granted one is going to be much faster than the other (we hope).

I've done this with explicit acknowledgements rather than TCP back-pressure because message producers were asynchronous WRT to the network thread. Alternatively you could set the byte limits on a counter or semaphore, and when Netty is done flushing them you can have it return the bytes to the semaphore. That provides a hard limit. I didn't do that because I didn't have promises with listeners and such.

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamWriter.java:
```diff
@@ -131,28 +139,95 @@ protected long totalSize()
+            {
+                logger.error("sending file block failed", future.cause());
+                channel.close();
+                session.sessionFailed(); /// ????????
```
???!!???

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamWriter.java:
```diff
@@ -131,28 +139,95 @@ protected long totalSize()
+        {
+            logger.info("JEB:: waited on condition but channel is still unwritable");
+            channel.close();
+            session.sessionFailed(); /// ?????
```
More ?????

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamWriter.java:
```diff
@@ -131,28 +139,95 @@ protected long totalSize()
+        return true;
+    }
+
+    private static class ChannelWritabiliityListsener extends ChannelDuplexHandler
```
Typo in name

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,401 @@
+    private static final int DEFAULT_CHANNEL_BUFFER_SIZE = 1 << 22;
+
+    // TODO:JEB should this be configurable? also, is this local to this stream session or global to all stream sessions or per-remote peer?
+    private static final int DEFAULT_MAX_PARALLEL_TRANSFERS = 4;
```
I think it should be global because it represents a resource limit. Ensuring fairness is an orthogonal issue. Definitely configurable I think the debate would be whether it's 1st class enough to deserve to be in cassandra.yml, and/or have JMX getter and setter etc.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,401 @@
+//        localAddress = new InetSocketAddress(FBUtilities.getLocalAddress(), 0);
+//        boolean secure = MessagingService.isEncryptedConnection(connectionAddress);
+//        remoteAddress = new InetSocketAddress(connectionAddress, secure ? DatabaseDescriptor.getSSLStoragePort() : DatabaseDescriptor.getStoragePort());
+        encryptionOptions = null;//secure ? DatabaseDescriptor.getServerEncryptionOptions() : null;
```
Is encryption something that needs to be done here still or will Netty take care of it?

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,401 @@
+
+    private final ExecutorService fileTransferExecutor;
+
+    private Channel controlMessageChannel;
```
Add a comment saying this is one of the inbound channels, and since we don't send anything on it we can use it to send control messages.

Is it possible to not have an inbound channel if we aren't receiving any data?

aweisberg replied:
I'm not sure what the story is with the control channel right, but there still isn't a comment?

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,401 @@
+        logger.trace("[Stream #{}] Scheduling keep-alive task with {}s period.", session.planId(), keepAlivePeriod);
+
+        // TODO:JEB make sure to shut this keep-alive future down
+//        ScheduledFuture<?> scheduledFuture = channel.eventLoop().scheduleAtFixedRate(new KeepAliveTask(session), 0, keepAlivePeriod, TimeUnit.SECONDS);
```
Did you move the keep alive functionality somewhere?

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,401 @@
+        return controlMessageChannel != null;
+    }
+
+    public void injectControlMessageChannel(Channel channel)
```
Is injecting the control message channel vs just creating it worth the mental overhead?

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,401 @@
+    }
+
+    @Override
+    public ChannelFuture sendMessage(StreamMessage message) //throws IOException
```
Are we calling sendMessage from the Netty thread and potentially blocking on creating a control message channel? Either Netty automatically re-enters it's event loop or we are getting lucky due to the round robin assignment of channels and not having a netty thread wait on itself.

I looked at the Netty code and it will detect deadlock and error, but not fix it by re-entering the event loop (which is madness anyways).

Can we end up with multiple control channels if two threads concurrently call sendMessage?

So generally speaking it's messy to ever allow Netty to block on anything and running large swaths of application code and callbacks is a recipe for doing that when most of it has been written in a blocking fashion. That's a driving motivator behind not having Netty do anything other than compression and encryption. When it comes to processing message payloads and invoking general application code we want to be careful. We really want a dumb pipe and to keep stuff off of the event loop.

For core C* it generally does some very light pre-processing and then hands verbs off to an execution stage. We should review the verbs and make sure that whatever the verbs are doing in the Netty thread is safe because we don't want the Netty thread to block on itself because it waited on a condition in Gossip or something that depends on a message it needs to deliver.

In VoltDB this was enforced by separating network and processing threads via a mailbox. The only code the network threads run that they don't own is the message serialization. You always need an actor on the other side to process the messages. You could override Mailbox.deliver to put processing in the network thread explicitly, but that was only done in one performance sensitive case where the handoff wasn't worth it.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java:
```diff
@@ -0,0 +1,236 @@
+                        continue;
+
+                    // StreamInitMessage & IncomingFileMessage each start new channels, and IMF needs a session to be established a priori
+                    if (message instanceof StreamInitMessage)
```
This is really a generic inbound message handler right? Are there potentially races here with multiple threads interacting with StreamResultFuture and StreamManager and the global state? Can they register different state and then end up with a different StreamResultFuture or session?

Ideally I think each stream session would get a dedicated control channel and a dedicated actor thread to orchestrate control messages. Completely decouple the whole thing and turn into a queue of events being processed serially. The only concurrency and parallelism is then files being transferred in parallel which is pretty clearly shared nothing other than resource limits which are generally pretty easy to reason about since there is no risk of deadlock.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java:
```diff
@@ -0,0 +1,236 @@
+                    }
+                    else if (message instanceof IncomingFileMessage)
+                    {
+                        // TODO: i'd be great to check if the session actually exists before slurping in the entire sstable
```
Oh hah yeah. Hiding it in the deserializer is a bit wonky.

aweisberg commented on src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,308 @@
+            if (buf.writableBytes() >= writeSize)
+                return true;
+
+            channel.writeAndFlush(buf);
```
Where is the error handling for this?

aweisberg replied:
Also same comment as before. Maybe you shouldn't flush until the caller calls flush? Since it's a stream buffering is expected.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java:
```diff
@@ -0,0 +1,236 @@
+            return;
+
+        state = State.CLOSED;
+        blockingIOThread.interrupt();
```
Never interrupt! If it's blocked on something close the thing it's blocked on or signal and that can wake it up.

It could throw interrupted exception anywhere and there is no guarantee it's where you expect or that it is handled correctly.

aweisberg commented on src/java/org/apache/cassandra/streaming/messages/IncomingFileMessage.java:
```diff
@@ -44,8 +49,16 @@ public IncomingFileMessage deserialize(ReadableByteChannel in, int version, Stre
         {
             DataInputPlus input = new DataInputStreamPlus(Channels.newInputStream(in));
             FileMessageHeader header = FileMessageHeader.serializer.deserialize(input, version);
+            session = StreamManager.instance.findSession(header.sender, header.planId, header.sessionIndex);
```
I'm not a huge fan of deserializers having this kind of stateful side effect.

aweisberg commented on src/java/org/apache/cassandra/streaming/messages/StreamMessage.java:
```diff
@@ -48,6 +48,11 @@ public static void serialize(StreamMessage message, DataOutputStreamPlus out, in
+        return 1 + message.type.outSerializer.serializedSize(message, version);
+    }
+
     public static StreamMessage deserialize(ReadableByteChannel in, int version, StreamSession session) throws IOException
```
Ah now I get it. Streaming originally used ReadableByteChannel which is inherently a copying API. Super weird code here though it's doing non-blocking reads?

aweisberg reviewed (commented):
This being my second pass through streaming I have a much better understanding of both these changes and streaming.

So I got a bit more opinionated. I am going to take another look at the core Netty patch and the use of streams there as well because I wonder whether that is also a candidate for leveraging the existing ByteBuffer stream code.

aweisberg commented on src/java/org/apache/cassandra/streaming/compress/CompressedStreamWriter.java:
```diff
@@ -79,12 +91,28 @@ public void write(DataOutputStreamPlus out) throws IOException
+                    if (!waitUntilWritable(channel))
+                        return;
+
+                    channel.writeAndFlush(outBuf).addListener(future -> {
```
Maybe you shouldn't flush until you are going to block waiting for writability? That could reduce signaling overhead.

jasobrown committed:
clarify and reduce the use of streams and channel on both the inbound and outbound stream paths. reuses the existing ByteBuffer-based Data*Plus classes, especially RebufferingInputStream

jasobrown committed:
- removed waitUntilWritable() from ByteBufDataOutputStreamPlus and replaced with a simple semaphore.
- cleaned up the close/cancel notification in StreamingInboundHander from the thread#interrupt() to using an atomic variable (polling it from the background thread).
- corrected the way ByteBufs and ByteBuffers are used in the stream classes

- restore CASSANDRA-7585 and bring back StreamingConnectionFactory, with modified versions of it's implementations and then inject the factory where we used to do  it.
- move stream versioning back into StreamMessage, instead of StreamSession (where I had moved it)
- minor clean ups and doc additions
- addressed TODO's and error handling

jasobrown committed:
extra logging and revert StreamReceiveTask

jasobrown committed:
fix a few minor issues that were unconvered by dtests

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -557,39 +557,46 @@ public void receive(StreamMessage message)
-                break;
-            default:
-                throw new AssertionError("unhandled StreamMessage type: " + message.getClass().getName());
+            logger.error("failed to process a streaming message properly", t);
```
Would be nice for these to have the address of the host this was from in it.

jasobrown committed:
slight refactoring and added tests for StreamCompressionSerializer and StreamingInboundHandler

jasobrown committed:
added utest for RebufferingByteBufDataInputPlus

jasobrown committed:
clean up remaining TODOs

aweisberg commented on src/java/org/apache/cassandra/net/async/NettyFactory.java:
```diff
@@ -113,6 +118,7 @@
                                         "MessagingService-NettyAcceptor-Threads", false);
         inboundGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "MessagingService-NettyInbound-Threads", false);
         outboundGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "MessagingService-NettyOutbound-Threads", true);
+        streamingGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "Streaming-Netty-Threads", false);
```
This is a benchmarking question, but there probably shouldn't be an inbound group and outbound group for regular messages either. Ideally it would be thread per core and load balanced, but Netty doesn't load balance so the right answer for how many event loop groups to have is going to have to come from measuring and might be workload specific.

I think we should put streaming in it's own group for now just to reduce the amount of owrk we have to do and the concern that streaming might block netty threads.

aweisberg replied:
Alternatively, maybe just use one larger group?

I'm not sure what the right answer is for IO ratio. I think that when all the Netty threads do is prepare messages you don't really need to mess with it. We don't actually have non-IO tasks they are all really IO tasks even non-IO tasks since they are serialization/deserialization.

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,178 @@
+        channelRateLimiter = new Semaphore(channel.config().getWriteBufferHighWaterMark(), true);
+    }
+
+    protected WritableByteChannel newDefaultChannel()
```
@override?

jasobrown replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/net/async/ByteBufDataOutputStreamPlus.java:
```diff
@@ -0,0 +1,178 @@
+ * and passing that to the super class as {@link #buffer}. When the {@link #buffer} is full or {@link #doFlush(int)}
+ * is invoked, the {@link #currentBuf} is published to the netty channel.
+ */
+public class ByteBufDataOutputStreamPlus extends BufferedDataOutputStreamPlus
```
I like how this presents the Netty channel as a stream. Seems like you could practically use it to front a Netty stream when serializing messages for intracluster messaging from inside the Netty thread. That would let you put multiple messages in one buffer before flushing.

jasobrown replied:
> That would let you put multiple messages in one buffer before flushing

Yup, I'm planning on doing that as a followup ticket :)

aweisberg commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -0,0 +1,225 @@
+            {
+                // consume any buffers from the queue *before* checking (and acting on) the closed status.
+                // that allows us to consume any last buffers before the end-of-stream.
+                currentBuf = queue.poll(1, TimeUnit.SECONDS);
```
I still think it's looks cleaner for this to be a take() and on close you would put Unpooled.EMPTY_BUFFER in as a sentinel indicating EOF/closed. That would remove the loop and also wake the reader up on close immediately.

I would +1 this though.

jasobrown replied:
Hmm, i like this better than what I have. I'm not a fan what I had with the `poll(long, TImeUnit)`, but it allowed you out of a blocking scenario. 

wrt to the `while` loop, I think i need to keep that in case there's a spurious wake up from `take()`, in which case it throws an `InterruptedException`. Does this makes sense, or am I mistaken?

aweisberg replied:
BlockingQueue.take() doesn't return spuriously.

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -314,7 +310,7 @@ public void addStreamRequest(String keyspace, Collection<Range<Token>> ranges, C
      * @param repairedAt the time the repair started.
      */
-    public synchronized void addTransferRanges(String keyspace, Collection<Range<Token>> ranges, Collection<String> columnFamilies, boolean flushTables, long repairedAt)
+    synchronized void addTransferRanges(String keyspace, Collection<Range<Token>> ranges, Collection<String> columnFamilies, boolean flushTables, long repairedAt)
```
Is this going to be called from a Netty thread? It's flushing memtables.

jasobrown replied:
damn, you are correct here. I'll try to figure out something <sigh>

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,354 @@
+    private static final Logger logger = LoggerFactory.getLogger(NettyStreamingMessageSender.class);
+
+    // TODO make me configurable??
+    private static final int DEFAULT_MAX_PARALLEL_TRANSFERS = FBUtilities.getAvailableProcessors();
```
Yeah definitely make it at least a system property just so it's not a recompile and redeploy if the number ends up being important.

jasobrown replied:
done

aweisberg replied:
So this seems to tie the fate of the number of transfers to an orthogonal property related to CPU resources. Transfers consume CPU, IO, and generate garbage, so the right number doesn't seem like it would necessarily be related to the amount of CPU resources available.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,354 @@
+    private static final int DEFAULT_MAX_PARALLEL_TRANSFERS = FBUtilities.getAvailableProcessors();
+
+    // a simple mechansim for allowing a degree of fairnes across multiple sessions
+    private static final Semaphore fileTransferSemaphore = new Semaphore(DEFAULT_MAX_PARALLEL_TRANSFERS);
```
Should you ask for a fair semaphore? I'm not super clear on how unfair it gets.

jasobrown replied:
good point. The `Semaphore` docs indicate you should always construct as fair unless you have a real need. for un-fair

jasobrown commented:
As to why there are two EventLoopGroups for messaging, well, I more or less
followed the pattern of the existing internode messaging. This was one of
the first things I wrote on this project, and haven't given it much thought
since (because I knew I'd becoming back to it someday :P).

I collapsed the two streaming EventLoopGroups because it really did seem
like overkill to have more than one for streaming. Now that I've moved all
disk IO out of the streaming threads, I could collapse those
EventLoopGroups into the internode messaging groups, and perhaps collapse
all EventLoopGroups down to one. At this point, I'd prefer to leave that as
a followup perf improvement ticket.

wrt netty's IO ratio, it's actually slightly mislabeled. it's real function
is to regulate how many channel outbound tasks to execute before checking
for more inbound tasks (i.e. buffers at the socket). As the current
outbound channels have near-zero inbound activity (as the channels are
still uni-directional), setting the IO ratio essentially tells the event
loop to not cap the number of outbound tasks before checking for more
inbound tasks - just process as many outbound tasks as you have.

On Wed, Jun 7, 2017 at 9:01 AM, aweisberg <notifications@github.com> wrote:

> *@aweisberg* commented on this pull request.
> ------------------------------
>
> In src/java/org/apache/cassandra/net/async/NettyFactory.java
> <https://github.com/jasobrown/cassandra/pull/2#discussion_r120669290>:
>
> > @@ -113,6 +118,7 @@
>                                          "MessagingService-NettyAcceptor-Threads", false);
>          inboundGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "MessagingService-NettyInbound-Threads", false);
>          outboundGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "MessagingService-NettyOutbound-Threads", true);
> +        streamingGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "Streaming-Netty-Threads", false);
>
> Alternatively, maybe just use one larger group?
>
> I'm not sure what the right answer is for IO ratio. I think that when all
> the Netty threads do is prepare messages you don't really need to mess with
> it. We don't actually have non-IO tasks they are all really IO tasks even
> non-IO tasks since they are serialization/deserialization.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jasobrown/cassandra/pull/2#discussion_r120669290>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABVYQXrD-Oly4IQJcdRSBIFMSwIAxKY0ks5sBslRgaJpZM4NTM4f>
> .
>

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -0,0 +1,354 @@
+        ChannelPipeline pipeline = channel.pipeline();
+        pipeline.addLast(NettyFactory.instance.streamingGroup, NettyFactory.INBOUND_STREAM_HANDLER_NAME, new StreamingInboundHandler(connectionId.remoteAddress(), protocolVersion, session));
+
+        int keepAlivePeriod = DatabaseDescriptor.getStreamingKeepAlivePeriod();
```
The keep alive plumbing still doesn't make any sense to me. 300 seconds as a default? But the lowest keep alive in the kernel config that is allowed is 2 hours. And the default is higher I think. 

And there is nothing using DatabaseDescriptor.getStreamingSocketTimeout()?

jasobrown replied:
tbh, I'm not a real fan of the keep-alive (introduced in CASSANDRA-11841, with additional background detail in CASSANDRA-8621), but it is a feature that is part of the server that I should support.

As in the CASSANDRA-8621 ticket, there' s *some* arguable justification, but, yeah, not a huge fan.

jasobrown replied:
> `DatabaseDescriptor.getStreamingSocketTimeout()`?

I believe that is a left over from CASSANDRA-11841 and was deprecated with 3.10. I'll remove for 4.0.

jasobrown commented:
As a followup though: it might be more advantageous to merge the internode
messaging EventLoopGroups with the client, native protocol EventLoopGroup
(in o.a.c.transport.Server and related classes) before addressing the
streaming EventLoopGroup, but that's a discussion for another ticket :)

On Wed, Jun 7, 2017 at 9:01 AM, aweisberg <notifications@github.com> wrote:

> *@aweisberg* commented on this pull request.
> ------------------------------
>
> In src/java/org/apache/cassandra/net/async/NettyFactory.java
> <https://github.com/jasobrown/cassandra/pull/2#discussion_r120669290>:
>
> > @@ -113,6 +118,7 @@
>                                          "MessagingService-NettyAcceptor-Threads", false);
>          inboundGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "MessagingService-NettyInbound-Threads", false);
>          outboundGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "MessagingService-NettyOutbound-Threads", true);
> +        streamingGroup = getEventLoopGroup(useEpoll, FBUtilities.getAvailableProcessors(), "Streaming-Netty-Threads", false);
>
> Alternatively, maybe just use one larger group?
>
> I'm not sure what the right answer is for IO ratio. I think that when all
> the Netty threads do is prepare messages you don't really need to mess with
> it. We don't actually have non-IO tasks they are all really IO tasks even
> non-IO tasks since they are serialization/deserialization.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jasobrown/cassandra/pull/2#discussion_r120669290>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABVYQXrD-Oly4IQJcdRSBIFMSwIAxKY0ks5sBslRgaJpZM4NTM4f>
> .
>

aweisberg commented on src/java/org/apache/cassandra/streaming/compress/CompressedInputStream.java:
```diff
@@ -114,88 +153,74 @@ private void decompressNextChunk() throws IOException
         if (crcCheckChance > 0d && crcCheckChance > ThreadLocalRandom.current().nextDouble())
         {
-            int checksum = (int) checksumType.of(compressed, 0, compressed.length - checksumBytes.length);
+            ByteBuffer crcBuf = compressed.duplicate();
```
Did we decide always validate or sometimes validate?

jasobrown replied:
For the current ticket, I"m going to leave the behavior as it is. If we want to change the behavior, we should open up a new ticket. In it's current state it's configurable by the operator, which is `ColumnFamilyStore#crcCheckChance`

aweisberg commented on src/java/org/apache/cassandra/streaming/compress/CompressedInputStream.java:
```diff
@@ -114,88 +153,74 @@ private void decompressNextChunk() throws IOException
-
-    @Override
-    public int read(byte[] b, int off, int len) throws IOException
+    private void decompress(ByteBuffer compressed) throws IOException
```
This needs more explanation as to why it sometimes uses the compressed buffer directly.

jasobrown replied:
comment added - basically it's because of CASSANDRA-10520

jasobrown committed:
add remote address to log message

aweisberg commented on src/java/org/apache/cassandra/streaming/StreamReader.java:
```diff
@@ -99,7 +97,8 @@ public SSTableMultiWriter read(ReadableByteChannel channel) throws IOException
 
-        TrackedInputStream in = new TrackedInputStream(new LZFInputStream(Channels.newInputStream(channel)));
+
+        TrackedDataInputPlus in = new TrackedDataInputPlus(new StreamCompressionInputStream(inputPlus, StreamMessage.CURRENT_VERSION));
```
So in a perfect world TrackeDataInputPlus for this use case would be per buffer not per primitive read since we don't really need that much precision. Save it for later though.

jasobrown replied:
Agreed. Adding to my stack of "future tickets" :)

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java:
```diff
@@ -0,0 +1,140 @@
+            compressed.limit(compressedLength).position(0);
+
+            // not sure if it's better to alloc a CompositeByteBuf and add the two buffers to it,
+            // or create another ByteBuf and copy the compressed data to it (after the headers)
```
You will use more memory since the buffer is sized to max compressed length. If that is fine you can just have compression output to an offset in the buffer leaving room for the header.

You need to do the math on worst case memory usage to decide if 2x overhead is acceptable.

aweisberg replied:
So one thing that I haven't thought of before is you don't technically need to allocate max length, you could allocate less than that. Say the uncompressed length + 10%, attempt to compress, and if that fails then reallocate. Max compressed length is ridiculously bad and I'm not sure it ever happens. If it does we can do something slower.

jasobrown replied:
Oops, this comment is a leftover from when this method was returning a netty `ByteBuf`.

Looking at the implementation of `net.jpountz.lz4.LZ4Utils#maxCompressedLength`, the value returned is `length + length / 255 + 16`. In the case of a 64k buffer (65535 bytes), the max length is 65808 bytes.

jasobrown replied:
...... and reading through `BufferPool`, I think there's a problem :( `BufferPool#maybeTakeFromPool` checks if the requested `size` is larger than `CHUNK_SIZE`, and if so, just calls `ByteBuffer#allocateDirect()`. And `CHUNK_SIZE` is a static final whose value is 64k.

So basically any buffer greater than 64k is not coming from a pool :-/ . As streaming is almost, by definition, going to use large buffers, I think I might need to ditch the use of `BufferPool`. wdyt?

jasobrown replied:
I made the executive decision to dump `BufferPool` as every time it 'misses' the cache (when a requested buffer is too large), it updates a bunch of metrics (which updates many volatile variables). Not so good for performance, not so good for metrics reporting.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java:
```diff
@@ -0,0 +1,140 @@
+            ByteBuffer uncompressed = null;
+            try
+            {
+                int readLength = ((ReadableByteChannel) in).read(compressed);
```
You don't need to make a read copy right? It's already in a buffer in the channel? Can you just point the decompressor at it? Or is there fragmentation?

aweisberg replied:
Any reaction to this?

jasobrown replied:
huh, maybe i didn't click "Comment" before, and I lost my comments :(

The reason why I'm copying here is that there's no guarantee the bytes you want to read (in the `in` paramaeter, basically a `RebufferingByteBufDataInputPlus`) are in one contiguous buffer. So you'll have to copy them into one buffer at some point if they are different buffers.

Thus, I thought implementing the `ReadableByteChannel` interface might be the least crappy solution for copying (on `RebufferingByteBufDataInputPlus`)

aweisberg commented on src/java/org/apache/cassandra/utils/memory/BufferPool.java:
```diff
@@ -61,7 +61,7 @@
 
     private static final Logger logger = LoggerFactory.getLogger(BufferPool.class);
-    private static final NoSpamLogger noSpamLogger = NoSpamLogger.getLogger(logger, 15L, TimeUnit.MINUTES);
+    private static final NoSpamLogger noSpamLogger = NoSpamLogger.getLogger(logger, 1, TimeUnit.SECONDS);
```
Is this change something we really want to land?

jasobrown replied:
derp, probably not. reverting ...

aweisberg commented on test/long/org/apache/cassandra/streaming/LongStreamingTest.java:
```diff
@@ -1,3 +1,4 @@
+
```
Whitespace?

jasobrown replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/tools/BulkLoadConnectionFactory.java:
```diff
@@ -43,24 +42,16 @@ public BulkLoadConnectionFactory(int storagePort, int secureStoragePort, Encrypt
+        return createConnection(connectionId, protocolVersion, encryptionOptions);
     }
 }
+
```
Whitespace?

jasobrown replied:
fixed

aweisberg reviewed (commented):
Looks pretty good. I guess the only thing really remaining to Slyvain's point is benchmarking. It would be great if you could take the knowledge you have today about benchmarking streaming and preserve it along with the code or scripts you made to do it for yourself.

jasobrown commented:
wrt `NettyFactory#createXXHashChecksum()`, i can't switch to a 64-bit checksum as the existing internode messaging is expecting a 32-bit checksum (based on the LZ4 library)

aweisberg commented:
But you could still use CRC32? That's going to be a quite a bit faster and it's going to avoid a copy.

jasobrown committed:
- clean up KeepAlive functionality
-- removed dead property {{streaming_socket_timeout_in_ms}} from {{Config}} and {{DatabaseDescriptor}}
-- properly send a {{KeepAliveMessage}} on each channel when it is not actively streaming a file.
-- refactorings for ease of testing
- change {{RebufferingByteBufDataInputPlus}} to use an empty buffer as a 'closed' indicator. This allowed replacing the {{queue#poll}} with {{queue#take}} and eliminate the {{while}} loop
- when a receiving node gets a {{PrepareSynMessage}}, asynchronously prepare the session as memtables might need to be flushed, and we don't want to have blocking disk IO on the netty event loop.
- minor formatting fixes, documenation updates, and add {{@Override}}
- added tests got {{NettyStreamingMessageSender}}
- factored out {{TestScheduledFuture}} to reuse it

jasobrown committed:
removed uses of BufferPool from streaming as it's best for small-medium size buffers (less than 64k)

jasobrown committed:
add more utests to NSMS, and minor bit of refactoring to support it; remove StreamMessage volatile field 'sent' as it's no longer used.

jasobrown committed:
protect against NPE

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java:
```diff
@@ -84,23 +91,23 @@ public StreamingInboundHandler(InetSocketAddress remoteAddress, int protocolVers
             buffers.append((ByteBuf) message);
         else
-            ctx.fireChannelRead(message);
+            ReferenceCountUtil.release(message);
```
You need

jasobrown commented:
In netty-land, it's safest to call `ReferenceCountUtil#release` on any
incoming message you aren't going to process. That way, `ByteBuf`s can be
properly released.

On Tue, Jun 13, 2017 at 11:11 AM, aweisberg <notifications@github.com>
wrote:

> *@aweisberg* commented on this pull request.
> ------------------------------
>
> In src/java/org/apache/cassandra/streaming/async/
> StreamingInboundHandler.java
> <https://github.com/jasobrown/cassandra/pull/2#discussion_r121754880>:
>
> >      {
>          if (!closed && message instanceof ByteBuf)
>              buffers.append((ByteBuf) message);
>          else
> -            ctx.fireChannelRead(message);
> +            ReferenceCountUtil.release(message);
>
> You need
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jasobrown/cassandra/pull/2#pullrequestreview-43808977>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVYQajOZIh0ibt1NtxCASW5-yxoXuAaks5sDtDDgaJpZM4NTM4f>
> .
>

jasobrown committed:
fix a few eclipse-warnings

aweisberg commented on src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java:
```diff
@@ -94,15 +87,15 @@ public ByteBuffer deserialize(LZ4FastDecompressor decompressor, DataInputPlus in
                 compressed.flip();
 
-                uncompressed = BufferPool.get(uncompressedLength);
+                uncompressed = ByteBuffer.allocateDirect(uncompressedLength);
```
What about Netty's allocator? That might be faster since it's supposed to be a jemalloc like implementation. allocateDirect will have to create a cleaner and cleaners can be kind of a shit show and are a global synchronization point. Maybe it's fine since only streaming will be doing this. I don't know.

jasobrown replied:
Good call. I've switched to use the `ByteBuffer`s I can get from a `ByteBuf`. Further, because of the way that netty give you the nio buffer, i don't run into the 'arrayOffset' bug in lz4 - which means I can ditch having an intermediate buffer to compress into/copy from :).

I also converted the API of this class, and fixed up the downstream callers.

aweisberg commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -93,10 +95,24 @@
+    private static final FastThreadLocal<Channel> threadLocalChannel = new FastThreadLocal<>();
+
+    /**
+     * A netty channel attribute used to indicate if a channel is currently transferring a file. This is primarily used
```
Yikes! Good catch.

aweisberg replied:
But now the issue is that's it's racey :-) Keep alive needs to hold a lock while checking the value and sending the message. Streaming needs to hold the lock while setting the value.

jasobrown replied:
Normally, yes, that would be completely true - that things are racey. However, since the `KeepAliveTask` executes as a task in the netty event loop (which is single-threaded for that channel), I'm guaranteed that a `KeepAliveTask` and a `FileStreamTask` will not execute concurrently in the same channel (and both try to update that channel attribute).

I will add some comments to that effect, though; that is good to have in there.

jasobrown replied:
We discussed offline, and I added comments, but I fixed a stupid bug. `KeepAliveTask` was writing to the control channel, not the channel on which it executed.

jasobrown committed:
refactor StreamCompressionSerializer to use ByteBufs, and clean up some related API use

jasobrown committed:
added more documentation to NettyStreamingMessageSender, and found a bug where KeepAliveTask was writing to the control channel, not it's own channel

jasobrown committed:
fixed native protocol TLS integration as some things changed between 4.0 and 4.1
